### ISSUE TITLE
Arithmetic overflow in set_len

### ISSUE THREAD

**Issue by yallop** [mirage/ocaml-cstruct#162]
`set_len` should
> `@raise Invalid_argument if [len] exceeds the size of the buffer.`

However:
```
# Cstruct.(set_len ((sub (create 10) 5 5)) max_int );;
- : Cstruct.t = {Cstruct.buffer = <abstr>; off = 5; len = 4611686018427387903}
```

More dramatic illustration:
```
# Cstruct.(blit (create 0xfffff) 0 (set_len ((sub (create 10) 5 5)) max_int ) 0 0xfffff);;
Segmentation fault
```

**Comment by avsm** [comment]
with #164 this should be now raising an exception in `check_bounds`

```
Cstruct.(set_len ((sub (create 10) 5 5)) max_int );;
Exception: Invalid_argument "Cstruct.set_len [5,5](10) 4611686018427387903".                                                                                           
```

Adding a test case...

**Comment by avsm** [comment]
And the second case:

```
Cstruct.(blit (create 0xfffff) 0 (set_len ((sub (create 10) 5 5)) max_int ) 0 0xfffff);;
Exception: Invalid_argument "Cstruct.set_len [5,5](10) 4611686018427387903"
```

**Comment by yallop** [comment]
Closing as fixed.  (The fix relies on the fact that adding two positive numbers cannot overflow and produce a positive number.)

**Comment by avsm** [comment]
Thanks @yallop! Fix will be in 3.1.0
